// azure-mobile-apps.data.memory/index.js
var promises = require('azure-mobile-apps.core/promises'),
    tables = {};

module.exports = function () {
    return {
        read: function (query) {
            // need to evaluate query against each item before returning
            return promises.resolved(values(query.getComponents().table.toLowerCase()));
        },
        update: function (table, item) {
            items(table)[item.id] = item;
            return promises.resolved(item);
        },
        insert: function (table, item) {
            items(table)[item.id] = item;
            return promises.resolved(item);
        },
        delete: function (table, id, version) {
            delete items(table)[id];
            return promises.resolved(1);
        }
    }

    function items(table) {
        table = table.toLowerCase();
        if(!tables[table])
            tables[table] = {};
        return tables[table];
    }

    function values(table) {
        table = table.toLowerCase();
        var table = items(table);
        return Object.keys(table).map(function (id) {
            return table[id];
        });
    }
}
