// azure-mobile-apps.express/tables/configurationFactory.js
var middlewareFactory = require('./middlewareFactory'),
    executeOperation = require('../middleware/executeOperation'),
    express = require('express');

module.exports = function () {
    // "default" configuration
    var router = express.Router(),
        configuration = {
            middleware: [],
            read: [],
            update: [],
            insert: [],
            delete: []
        };
    
    var factory = function () {
        return middlewareFactory(configuration, router);
    };
    
    factory.execute = router;
    factory.operation = executeOperation;
    factory.use = attachMiddleware('middleware');
    factory.use.read = attachMiddleware('read');
    factory.use.update = attachMiddleware('update');
    factory.use.insert = attachMiddleware('insert');
    factory.use.delete = attachMiddleware('delete');
    
    return factory;
    
    function attachMiddleware(property) {
        return function (middleware) {
            Array.prototype.push.apply(configuration[property], arguments);
        }
    }
}
