// azure-mobile-apps.express/test/configuration/tables/raw.js
var tables = require('azure-mobile-apps.express').tables;
 
module.exports = {
    softDelete: true,
    undelete: true,
    
    // the middleware property executes supplied middleware for all operations - tables.execute represents the middleware defined for each operation
    middleware: [tables.execute, log],
    
    // the executeRead function generates middleware that will execute the provided function with a (query, context) signature
    read: tables.operation(function (query, context) {
        return context.execute().then(function (results) {
            return results.forEach(function (item) {
                item.loaded = new Date();
            });
        });
    }),
    
    // we can pass an array of middleware to execute
    insert: [authorise, tables.operation(function (item, context) {
            return context.execute().then(function () {
                log.debug('New record inserted - ', JSON.stringify(item));
            });
        })],
    
    // not passing a function to executeXxx will execute the operation with default behavior
    update: [authorise, tables.operation(), log]
    
    // omitting an operation altogether will result in default behavior
    //delete: [tables.operation()]
}

function log(req, res, next) { }
function authorise(req, res, next) { }
