// azure-mobile-apps.express/tableConfigurationFactory.js
var loader = require('azure-mobile-apps.core/configuration/loader');

module.exports = function () {
    // "default" configuration
    var configuration = {
        middleware: [this.execute],
        read: [this.operation],
        update: [this.operation],
        insert: [this.operation],
        delete: [this.operation]
    };
    
    var factory = function () {
        return configuration;
    };
    
    // middleware to be added to indicate when table operation should execute as part of global table flow
    factory.execute = function (req, res, next) {

    };

    // middleware to be added to indicate when table operation should execute as part of specific table operation flow
    factory.operation = function (req, res, next) {

    };
    
    factory.use = attachMiddleware('middleware');
    factory.use.read = attachMiddleware('read');
    factory.use.update = attachMiddleware('update');
    factory.use.insert = attachMiddleware('insert');
    factory.use.delete = attachMiddleware('delete');
    
    return factory;

    function attachMiddleware(property) {
        return function (middleware) {
            configuration[property] = Array.prototype.slice.call(arguments);
        }
    }
}

// static versions of the execute and operation functions for use in raw table configurations
module.exports.execute = function (userCode) {
    
}

module.exports.operation = function (userCode) {

}
