// azure-mobile-apps.data.sql/dynamicSchema.js
var statements = require('./statements'),
    execute = require('./execute'),
    promises = require('azure-mobile-apps.core/promises'),
    log = require('azure-mobile-apps.core/logger');

var errorCodes = {
    InvalidColumnName: 207,
    InvalidObjectName: 208
}

module.exports = function (config) {
    return {
        execute: function (table, statement, item, attempt) {
            attempt = attempt || 1;
            if (attempt > 3)
                throw new Error('Failed to update schema');

            return execute(config, statement)
                .catch(function (err) {
                    return handleError(err).then(function () {
                        return executeWithRetry(table, statement, item, attempt + 1);
                    });
                });

            function handleError(err) {
                if(err.number === errorCodes.InvalidObjectName)
                    return createTable();
                if(err.number === errorCodes.InvalidColumnName)
                    return updateSchema();
                return promises.rejected(err);
            }

            function createTable() {
                log.info('Creating table ' + table.name);
                return execute(config, statements.createTable(table));
            }

            function updateSchema() {
                log.info('Updating schema for table ' + table.name);
                return execute(config, statements.getColumns(table)).then(function (columns) {
                    return execute(config, updateSchema(table, columns, item));
                });
            }
        }
    };
};
