// azure-mobile-apps.data.sql/dynamicSchema.js
var statements = require('./statements'),
    execute = require('./execute'),
    promises = require('azure-mobile-apps.core/promises'),
    log = require('azure-mobile-apps.core/logger');

var errorCodes = {
    InvalidColumnName: 207,
    InvalidObjectName: 208
}

module.exports = function (config) {
    var api = {
        execute: function (table, statement, item, attempt) {
            attempt = attempt || 1;

            return execute(config, statement)
                .catch(function (err) {
                    return handleError(err).then(function () {
                        return api.execute(table, statement, item, attempt + 1);
                    });
                });

            function handleError(err) {
                if (attempt >= 3)
                    return promises.rejected(err);

                if(err.number === errorCodes.InvalidObjectName)
                    return createTable();
                if(err.number === errorCodes.InvalidColumnName)
                    return updateSchema();
                return promises.rejected(err);
            }

            function createTable() {
                log.info('Creating table ' + table.name);
                return execute(config, statements.createTable(table));
            }

            function updateSchema() {
                log.info('Updating schema for table ' + table.name);
                return execute(config, statements.getColumns(table)).then(function (columns) {
                    return execute(config, statements.updateSchema(table, columns, item));
                });
            }
        },
        read: function (table, statement) {
            return execute(config, statement)
                .catch(function (err) {
                    // if dynamic schema is enabled and the error is invalid column, it is likely that the schema has not been
                    // updated to include the column. V1 behavior is to return an empty array, maintain this behavior.
                    if((err.number === errorCodes.InvalidColumnName || err.number === errorCodes.InvalidObjectName)
                        && (table.dynamicSchema === undefined || table.dynamicSchema))
                        return [];
                    throw err;
                });
        }
    };

    return api;
};
